### DISCUSSION TITLE
Can nodes and edges have additional attributes?

### DISCUSSION THREAD

**Discussion by cooknl** [arun1729/cog#31] [Q&A] [answered]
I'd like to include additional attributes to both nodes and edges, preferably arbitrary, but common across all "like" objects

As an example, imagine people nodes
Attributes could include contact information, biographical information, b etc

Imagine a "x called y" edge
Attributes could include datetime of call, call notes, duration, etc

**Comment by arun1729** [comment, marked as answer]
This is actually a good idea but currently nodes are limited to just strings. Generally in a triple-store database, everything is represented in the form of subject-predicate-object triples. So, in the case of "x called y", the attributes can be represented like this:
x called call-id
call-id name y
call-id datetime <datetime>
call-id callnotes <call notes>
call-id duration <duration>
... etc

call-id (can be any unique set of characters) links all the attributes together.

But I do think it would really nice to have nodes be more than just strings and be able to store attributes (a dictionary of sorts). One advantage is that attributes do not need to be made into graphs and make the "main" graph more complex than necessary. Graph would then be more "cleaner" and represent only the main relationship. This needs a bit more thought but could be added to the database as a feature. Thanks for bringing this up!
